crates/matrix-sdk/src/notification_settings/rules.rs

```diff
@@ -0,0 +1,1091 @@
+    /// actions to be performed on the user's account data.
+    ///
+    /// # Arguments
+    ///
+    /// * `rules` - A list of rules to delete
+    /// * `exceptions` - A list of rules to ignore
+    pub(crate) fn delete_rules(
+        &mut self,
+        rules: &[(RuleKind, String)],
+        exceptions: &[(RuleKind, String)],
+    ) -> Result<Vec<Command>, RemovePushRuleError> {
+        let mut commands: Vec<Command> = vec![];
+
+        // Remove exceptions from the rules to delete
+        let to_remove = BTreeSet::from_iter(exceptions);
+        let mut rules = rules.to_vec();
+        rules.retain(|e| !to_remove.contains(e));
+
+        for (rule_kind, rule_id) in rules {
+            self.ruleset.remove(rule_kind.clone(), rule_id.clone())?;
+            commands.push(Command::DeletePushRule {
+                scope: RuleScope::Global,
+                kind: rule_kind.clone(),
+                rule_id: rule_id.clone(),
+            })
+        }
+        Ok(commands)
+    }
+
+    /// Sets whether a rule is enabled and returns a list of `Command`
+    /// describing the actions to be performed on the user's account data.
+    pub(crate) fn set_enabled(
+        &mut self,
+        kind: RuleKind,
+        rule_id: &str,
+        enabled: bool,
+    ) -> Result<Vec<Command>, NotificationSettingsError> {
+        if rule_id == PredefinedOverrideRuleId::IsRoomMention.as_str() {
+            // Handle specific case for `PredefinedOverrideRuleId::IsRoomMention`
+            self.set_room_mention_enabled(enabled)
+        } else if rule_id == PredefinedOverrideRuleId::IsUserMention.as_str() {
+            // Handle specific case for `PredefinedOverrideRuleId::IsUserMention`
+            self.set_user_mention_enabled(enabled)
+        } else {
+            let mut commands = vec![];
+            self.set_rule_enabled(kind, rule_id, enabled, &mut commands)?;
+            Ok(commands)
+        }
+    }
+
+    /// Helper function to enable or disable a rule and update a list of
+    /// commands to be performed on the user's account data to reflect the
+    /// change.
+    fn set_rule_enabled(
+        &mut self,
+        kind: RuleKind,
+        rule_id: &str,
+        enabled: bool,
+        commands: &mut Vec<Command>,
+    ) -> Result<(), NotificationSettingsError> {
+        self.ruleset.set_enabled(kind.clone(), rule_id, enabled)?;
+        commands.push(Command::SetPushRuleEnabled {
+            scope: RuleScope::Global,
+            kind,
+            rule_id: rule_id.to_owned(),
+            enabled,
+        });
+
+        Ok(())
+    }
+
+    /// Set whether the `IsUserMention` rule is enabled and returns a list of
+    /// `Command` describing the actions to be performed on the user's account
+    /// data.
+    fn set_user_mention_enabled(
+        &mut self,
+        enabled: bool,
+    ) -> Result<Vec<Command>, NotificationSettingsError> {
+        let mut commands: Vec<Command> = vec![];
```

Comment:
Pretty sure you can leave the type annotation here, type inference goes backward too! rustc will see a `Vec<Command>` is expected in `set_rule_enabled` and infer the correct type.

```suggestion
        let mut commands = vec![];
```